src/v/raft/offset_translator.cc

```diff
@@ -0,0 +1,506 @@
+ * by the Apache License, Version 2.0
+ */
+
+#include "raft/offset_translator.h"
+
+#include "model/adl_serde.h"
+#include "raft/consensus_utils.h"
+#include "reflection/adl.h"
+#include "serde/serde.h"
+#include "storage/api.h"
+#include "vlog.h"
+
+namespace raft {
+
+static ss::logger logger{"offset_translator"};
+
+offset_translator::offset_translator(
+  std::vector<model::record_batch_type> filtered_types,
+  raft::group_id group,
+  model::ntp ntp,
+  storage::api& storage_api)
+  : _filtered_types(std::move(filtered_types))
+  , _group(group)
+  , _ntp(ntp)
+  , _logger(logger, ssx::sformat("ntp: {}", _ntp))
+  , _storage_api(storage_api) {}
+
+int64_t offset_translator::delta(model::offset o) const {
+    if (_filtered_types.empty()) {
+        return 0;
+    }
+
+    auto it = _offset2delta.lower_bound(o);
+    if (it == _offset2delta.begin()) {
+        throw std::runtime_error{_logger.format(
+          "log offset {} is outside the translation range (starting at {})",
+          o,
+          _offset2delta.begin()->first)};
+    }
+    return std::prev(it)->second;
+}
+
+model::offset offset_translator::from_log_offset(model::offset o) const {
+    auto d = delta(o);
+    return model::offset(o - d);
+}
+
+model::offset offset_translator::to_log_offset(
+  model::offset data_offset, model::offset hint) const {
+    if (_filtered_types.empty()) {
+        return data_offset;
+    }
+
+    if (data_offset == model::offset::max()) {
+        return data_offset;
+    }
+
+    vassert(
+      !_offset2delta.empty(), "ntp {}: offsets map shouldn't be empty", _ntp);
+
+    model::offset min_data_offset
+      = details::next_offset(_offset2delta.begin()->first)
+        - model::offset(_offset2delta.begin()->second);
+    vassert(
+      data_offset >= min_data_offset,
+      "ntp {}: data offset {} must be inside translation range (starting at "
+      "{})",
+      _ntp,
+      data_offset,
+      min_data_offset);
+
+    // We iterate over the intervals (beginning exclusive, end inclusive)
+    // with constant delta, starting at the interval containing
+    // log offset equal to `data_offset` (because log offset is at least as
+    // big as data offset) and stopping when we find the interval where
+    // given data offset is achievable.
+    auto interval_end_it = _offset2delta.lower_bound(
+      std::max(hint, data_offset));
+    auto delta = std::prev(interval_end_it)->second;
+
+    while (interval_end_it != _offset2delta.end()) {
+        // last data record in this interval is at offset
+        // interval_end_it->first - 1
+        model::offset max_do_this_interval
+          = details::prev_offset(interval_end_it->first) - model::offset(delta);
+        if (max_do_this_interval >= data_offset) {
+            break;
+        }
+
+        delta = interval_end_it->second;
+        ++interval_end_it;
+    }
+
+    return data_offset + model::offset(delta);
+}
+
+void offset_translator::process(
```

Comment:
do we need to pass `base_offset` ? we have it in `model::record_batch` already

Reply:
I added it because appended batch has zero base offset before it is passed to the appender. But we can add to offset translator after we append so that `base_offset` is not needed.

Reply:
I think that it may be beneficial to add it after append but before making batches visible in Kafka layer. This way we wouldn't have to duplicate offset assigning logic in translator.